server/routes/post.route.js

```diff
@@ -28,11 +28,20 @@ router.route('/:postId/comment')
     , commentCtrl.create);
 
 router.route('/:postId/upvote')
-  .post(expressJwt({ secret: config.jwtSecret }), voteCtrl.upvote);
+  .post(expressJwt({ secret: config.jwtSecret })
+  , voteCtrl.movePostToEntity
```

Comment:
You shouldn't define the step in middleware. The the controller should detail the flow.

Reply:
I'm not exactly following here. What is it that you are suggesting? 

The reasoning behind this was that every client would from now on pass an entity instead of a post so the vote controller should not be aware of each type.

Reply:
I think the issue is because the controller is so tied to MVC, but again this is just like my OO example I sent you.

1. For future reference, something like this should mostly be in one function `vote.controller.upVote` w.e. In that case, the Vote controller would determine the type - which is the correct first step.
2. What should do is switch the responsibility. The Vote controller should be not take the responsibility of generic voting. It should be a service that computes.. so
3. You need to have a post.controller.upvote and a comment.controller.upvote. Those functions should handle calling the vote logic and passing in the "entity". And, because it is a clean responsibility, you can use the post router and the comment router to load the "entities". 

It remove type checking, it pushes responsibility onto the entities (speaking in architecture terms) and abstracts to a vote service to handle shared logic. Then you also have the recommender service - used by the post entity for now.